src/cargo/ops/registry.rs

```diff
@@ -55,6 +60,45 @@ pub fn publish(manifest_path: &Path,
     Ok(())
 }
 
+fn check_git_cleanliness(repo: &git2::Repository) -> CargoResult<()>{
+    let mut opts = git2::StatusOptions::new();
+    opts.include_untracked(true).recurse_untracked_dirs(true);
+    let status = try!(repo.statuses(Some(&mut opts)));
+    match status.len(){
+        0 => {
+            Ok(())
+        },
+        num @ _ => {
+            let files:Vec<String> = status.iter().map(|entry| {
+                            let file = entry.index_to_workdir()
+                                 .unwrap()
+                                 .old_file()
+                                 .path()
+                                 .unwrap()
+                                 .display();
+                            format!("{}", file)
+                        }).collect();
+
+            Err(human(format!("{} uncommited or untacked files \
+        that need to be addressed before publishing\n{}", num, files.join("\n"))))
```

Comment:
Can this message also mention the `--allow-untracked` option to `cargo publish` as a possible escape hatch?